lib/None_Top_Bot.thy

```diff
@@ -0,0 +1,244 @@
+lemma none_bot_Sup:
+  "none_bot f opt = (Sup {f x |x. opt = Some x} :: 'a :: complete_lattice)"
+  by (cases opt) auto
+
+lemma none_bot_if:
+  "none_bot f opt = (if opt = None then bot else f (the opt))"
+  by (cases opt) auto
+
+lemma none_bot_bot[simp]:
+  "none_bot bot = bot"
+  by (rule ext) (simp split: none_bot_splits)
+
+
+section \<open>Bool/fun lemmas for @{const none_bot}\<close>
+
+lemma none_bot_bool_ex:
+  "none_bot f opt = (\<exists>x. opt = Some x \<and> f x)"
+  by (auto simp: none_bot_Sup)
+
+lemma none_bot_boolI:
+  "\<exists>x. opt = Some x \<and> f x \<Longrightarrow> none_top f opt"
+  by (auto simp: none_bot_bool_ex)
+
+lemma none_bot_boolD:
+  "none_bot f opt \<Longrightarrow> \<exists>x. opt = Some x \<and> f x"
+  by (cases opt) auto
+
+lemma none_bot_boolE:
+  "\<lbrakk> none_bot f opt; \<And>x. \<lbrakk> opt = Some x; f x \<rbrakk> \<Longrightarrow> P \<rbrakk> \<Longrightarrow> P"
+  by (cases opt) auto
+
+(* As for none_top_bot, it would be unusual to see "top" for bool. Only adding the lemma here for
+   completeness *)
+lemma none_bot_top_neq_None[simp]:
+  "none_bot top opt = (opt \<noteq> None)"
+  by (cases opt) auto
+
+lemma none_bot_True_bool_neq_None[simp]:
+  "none_bot \<top> opt = (opt \<noteq> None)"
+  by (cases opt) auto
+
+lemma none_bot_False_bool[simp]:
+  "none_bot \<bottom> = \<bottom>"
+  by (rule ext) (simp split: none_bot_split)
+
+lemma none_bot_False_boolE[dest!]:
+  "none_bot \<bottom> opt \<Longrightarrow> False"
+  by simp
+
+lemma none_bot_fun_ex:
+  "none_bot f opt s = (\<exists>x. opt = Some x \<and> f x s)"
+  by (cases opt) auto
+
+lemma none_bot_funI:
+  "\<exists>x. opt = Some x \<and> f x s \<Longrightarrow> none_bot f opt s"
+  by (simp add: none_bot_fun_ex)
+
+lemma none_bot_funD:
+  "none_bot f opt s \<Longrightarrow> \<exists>x. opt = Some x \<and> f x s"
+  by (cases opt) auto
+
+lemma none_bot_funE:
+  "\<lbrakk> none_bot f opt s; \<And>x. \<lbrakk> opt = Some x; f x s \<rbrakk> \<Longrightarrow> P \<rbrakk> \<Longrightarrow> P"
+  by (cases opt) auto
+
+lemma none_bot_True_fun_neq_None[simp]:
+  "none_bot \<top>\<top> opt s = (opt \<noteq> None)"
+  by (cases opt) auto
+
+lemma none_bot_False_fun[simp]:
+  "none_bot \<bottom>\<bottom> = \<bottom>\<bottom>"
+  by (rule ext) (simp split: none_bot_split)
+
+lemma none_bot_False_funE[dest!]:
+  "none_bot \<bottom>\<bottom> opt s \<Longrightarrow> False"
+  by simp
+
+section \<open>Automation setup and short-hand names\<close>
+
+lemmas none_topI[intro!] = none_top_boolI none_top_funI
+lemmas none_topD = none_top_boolD none_top_funD
+lemmas none_topE = none_top_boolE none_top_funE
+lemmas none_top_all = none_top_bool_all none_top_fun_all
+lemmas none_top_case = none_top_bool_cases none_top_fun_cases
+
+lemmas none_botI = none_bot_boolI none_bot_funI
+lemmas none_botD = none_bot_boolD none_bot_funD
+lemmas none_botE[elim!] = none_bot_boolE none_bot_funE
+lemmas none_bot_ex = none_bot_bool_ex none_bot_fun_ex
```

Comment:
They're duals also in the sense of automation. For `top`, the intro rule is nice and safe and something you would always need to do when proving `none_top`, but the elim rule introduces cases (it's also safe, but splitting might lead to exponential blow-up if you have many of these in the goal). For `bot`, it's the other way around: the elim rule is the safe one in terms of logic, automation, and intention, but would introduce an existential in the intro case, which one might want to avoid.